chain/client/src/sync/handler.rs

```diff
@@ -0,0 +1,367 @@
+                // we just need to download more headers.)
+                return Ok(false);
+            }
+            if epoch_sync_boundary_block_header.height() + chain.transaction_validity_period()
+                > current_epoch_start
+            {
+                // We also do not want to state sync, if by doing so we would not have enough headers to
+                // perform transaction validity checks. Again, epoch sync should have picked an old
+                // enough epoch to ensure that we would have enough headers if we just continued with
+                // header sync.
+                return Ok(false);
+            }
+        }
+
+        let block_sync_result = self.block_sync.run(
+            &mut self.sync_status,
+            &chain,
+            highest_height,
+            highest_height_peers,
+            self.config.sync_max_block_requests,
+        )?;
+        Ok(block_sync_result)
+    }
+
+    /// Verifies if the node possesses the given block. If the block is absent,
+    /// the node should request it from peers.
+    ///
+    /// the return value is a tuple (request_block, have_block)
+    ///
+    /// the return value (false, false) means that the node already requested
+    /// the block but hasn't received it yet
+    fn sync_block_status(
+        &self,
+        chain: &Chain,
+        sync_hash: &CryptoHash,
+        block_hash: &CryptoHash,
+        now: Utc,
+    ) -> Result<(bool, bool), near_chain::Error> {
+        // The sync hash block is saved as an orphan. The other blocks are saved
+        // as regular blocks. Check if block exists depending on that.
+        let block_exists = if sync_hash == block_hash {
+            chain.is_orphan(block_hash)
+        } else {
+            chain.block_exists(block_hash)?
+        };
+
+        if block_exists {
+            return Ok((false, true));
+        }
+        let timeout = self.config.state_sync_external_timeout;
+        let timeout = near_async::time::Duration::try_from(timeout);
+        let timeout = timeout.unwrap();
+
+        let Some(last_time) = self.last_time_sync_block_requested.get(block_hash) else {
+            return Ok((true, false));
+        };
+
+        if (now - *last_time) >= timeout {
+            tracing::error!(
+                target: "sync",
+                %block_hash,
+                ?timeout,
+                "State sync: block request timed out"
+            );
+            Ok((true, false))
+        } else {
+            Ok((false, false))
+        }
+    }
+
+    /// Checks if the sync blocks are available and requests them if needed.
+    /// Returns true if all the blocks are available.
+    fn request_sync_blocks(
+        &mut self,
+        chain: &Chain,
+        block_header: BlockHeader,
+        highest_height_peers: &[HighestHeightPeerInfo],
+    ) -> Result<Vec<(CryptoHash, PeerId)>, near_chain::Error> {
+        let now = self.clock.now_utc();
+
+        let mut have_all = true;
```

Comment:
nit: this used to be returned so the caller can return from `handle_sync_needed()`, but now it's just logged at the bottom, so could just get rid of that and log whatever you want, maybe blocks_to_request.len() or something